### ISSUE TITLE
Memory bank

### ISSUE THREAD

**Issue by mshooter** [gabry1998/Self-Supervised-Anomaly-Detection#2]
Hi great work! 

I was wondering if you could elaborate on the memory bank? 
I am new to this and I see that you just update the memory bank per iteration, if I am correct? 
What is the difference between your method and PatchCore? 

And could you explain the patch level inference.

Thank you!

**Comment by gabry1998** [comment]
Hello Moira! Apologies for the delayed response.

If I recall correctly (I did the project last year, ahah), the purpose of the memory bank is to store embedding vectors related to predicted images without anomalies. This is done to save time in the inference phase. Having a ready set of non-anomalous elements allows you to immediately produce embedding vectors of images with REAL anomalies and compare them with the elements in the memory bank (using cosine similarity), as outlined in the readme.

For the memory bank, I didn't reinvent the wheel 😂. The logic is very similar to patchcore, but I didn't make use of the coreset subsampling phase. I simply update the elements with new data and discard the oldest ones if the bank size exceeds a certain value.

The patch level briefly consists of splitting the images into small square sub-images. For each sub-image, an anomaly score is evaluated in a similar way as the image-level (in this case, the memory bank elements consist of embedding vectors representing anomaly-free patches). After calculating the anomaly score for each patch of the image, I use bilinear interpolation to upscale the image to the original size.

I hope I have explained everything clearly 😅.
